infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java

```diff
@@ -0,0 +1,197 @@
+                "be flushed.");
+        for (CorfuMsg message : this.messages) {
+            ctx.writeAndFlush(message);
+        }
+
+        // Remove this handler from the pipeline; handshake is completed.
+        log.info("channelRead: Removing handshake handler from pipeline.");
+        ctx.pipeline().remove(this);
+        this.fireHandshakeSucceeded();
+    }
+
+    @Override
+    public void channelActive(final ChannelHandlerContext ctx) throws Exception {
+        log.info("channelActive: Incoming connection established from: " +
+                ctx.channel().remoteAddress() + " Start Read Timeout.");
+        ctx.pipeline().addBefore(ctx.name(), "readTimeoutHandler",
+                new ReadTimeoutHandler(10));
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        log.debug("channelInactive: Channel closed.");
+        if (!this.handshakeComplete.get()) {
+            this.fireHandshakeFailed(ctx);
+        }
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause)
+            throws Exception {
+        log.error("exceptionCaught: Exception caught: " + cause.getMessage());
+
+        if (cause instanceof ReadTimeoutException) {
+            // Read timeout: no inbound traffic detected in a period of time.
+            if (handshakeFailed.get()) {
+                log.debug("exceptionCaught: Handshake timeout checker: already failed.");
+                return;
+            }
+
+            if (!handshakeComplete.get()) {
+                log.error("exceptionCaught: Handshake timeout checker: timed out. Close Connection.");
+                handshakeFailed.set(true);
+                ctx.channel().close();
+            } else {
+                log.debug("exceptionCaught: Handshake timeout " +
+                        "checker: discarded (handshake OK)");
+            }
+        } else {
+            super.exceptionCaught(ctx, cause);
+        }
+
+        if (ctx.channel().isActive()) {
+            // Closing the channel will trigger handshake failure.
+            ctx.channel().close();
+        } else {
+            // Channel did not open, fire handshake failure.
+            this.fireHandshakeFailed(ctx);
+        }
+    }
+
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+            throws Exception {
+        if (this.handshakeFailed.get()) {
+            // If handshake failed, discard messages.
+            return;
+        }
+
+        if (this.handshakeComplete.get()) {
+            log.debug("write: Handshake already completed, not appending corfu message to queue");
+            super.write(ctx, msg, promise);
+        } else {
+            this.messages.offer((CorfuMsg) msg);
+        }
+    }
+
+
+    private void fireHandshakeFailed(ChannelHandlerContext ctx) {
```

Comment:
do we need 2 variables "handshakeComplete and handshakeFailed" to maintain state ?

Reply:
I believe we should use 2. The fact that handshake hasn't failed (default) does not imply it has completed. Only if it has completed we might forward incoming messages in the pipeline when received in the window of handshake validation when handler has not been removed yet from the pipeline.

Reply:
The variables you use are java objects. These variables have a derived state "initialized/null" and you can use that for the signal you are looking for. I am ok with your approach and do not have a strong opinion. If 2 variables make it easier for you, I would suggest to couple them together in a state object so that they mutate together and there is little chance of making mistakes where we set one and forgot the other. The code that you have does not have any issues, the logic looks good , this is more of suggestion for future proofing.